llvm/include/llvm/ABI/Types.h

```diff
@@ -0,0 +1,538 @@
+};
+
+class ComplexType : public Type {
+public:
+  ComplexType(const Type *ElementType, uint64_t SizeInBits, Align Alignment)
+      : Type(TypeKind::Complex, TypeSize::getFixed(SizeInBits), Alignment),
+        ElementType(ElementType) {}
+
+  const Type *getElementType() const { return ElementType; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Complex;
+  }
+
+private:
+  const Type *ElementType;
+};
+
+class IntegerType : public Type {
+private:
+  bool IsSigned;
+  bool IsBoolean;
+  bool IsBitInt;
+  uint64_t ABISizeInBits;
+
+public:
+  IntegerType(uint64_t BitWidth, uint64_t ABIWidth, Align Align, bool Signed,
+              bool IsBool = false, bool BitInt = false)
+      : Type(TypeKind::Integer, TypeSize::getFixed(BitWidth), Align),
+        IsSigned(Signed), IsBoolean(IsBool), IsBitInt(BitInt),
+        ABISizeInBits(ABIWidth) {}
+
+  bool isSigned() const { return IsSigned; }
+  bool isBool() const { return IsBoolean; }
+  bool isBitInt() const { return IsBitInt; }
+  uint64_t getLogicalBitWidth() const {
+    return getSizeInBits().getFixedValue();
+  }
+  uint64_t getABISizeInBits() const { return ABISizeInBits; }
+  bool isABIWidened() const { return getLogicalBitWidth() != ABISizeInBits; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Integer;
+  }
+};
+
+class FloatType : public Type {
+private:
+  const fltSemantics *Semantics;
+
+public:
+  FloatType(const fltSemantics &FloatSemantics, Align Align)
+      : Type(TypeKind::Float,
+             TypeSize::getFixed(APFloat::getSizeInBits(FloatSemantics)), Align),
+        Semantics(&FloatSemantics) {}
+
+  const fltSemantics *getSemantics() const { return Semantics; }
+  static bool classof(const Type *T) { return T->getKind() == TypeKind::Float; }
+};
+
+class PointerLikeType : public Type {
+protected:
+  unsigned AddrSpace;
+  PointerLikeType(TypeKind K, TypeSize Size, Align Align, unsigned AS = 0)
+      : Type(K, Size, Align), AddrSpace(AS) {}
+
+public:
+  virtual ~PointerLikeType() = default;
+  unsigned getAddrSpace() const { return AddrSpace; }
+  virtual bool isMemberPointer() const = 0;
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Pointer ||
+           T->getKind() == TypeKind::MemberPointer;
+  }
+};
+
+class PointerType : public PointerLikeType {
+public:
+  PointerType(uint64_t Size, Align Align, unsigned AddressSpace = 0)
+      : PointerLikeType(TypeKind::Pointer, TypeSize::getFixed(Size), Align,
+                        AddressSpace) {}
+
+  bool isMemberPointer() const override { return false; }
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Pointer;
+  }
+};
+
+class MemberPointerType : public PointerLikeType {
+private:
+  bool IsFunctionPointer;
+  bool Has64BitPointers;
+
+public:
+  MemberPointerType(bool IsFunctionPointer, bool Has64BitPointers,
```

Comment:
I don't think Has64BitPointers should be a property of MemberPointerType, it's a property of the target.